File: benchmarks/cpp/binary_trees.cpp
#include <cstdio>
struct Node { int v; Node *l, *r; };
Node* make_tree(int depth) {
    Node* n = new Node{1, nullptr, nullptr};
    if (depth > 0) { n->l = make_tree(depth-1); n->r = make_tree(depth-1); }
    return n;
}
long long check_tree(Node* n) {
    if (!n->l) return 1;
    return 1 + check_tree(n->l) + check_tree(n->r);
}
void free_tree(Node* n) { if (!n) return; free_tree(n->l); free_tree(n->r); delete n; }
int main() {
    int max_depth = 12;
    long long total = 0;
    for (int d = 4; d <= max_depth; d += 2) {
        int iters = 1 << (max_depth - d + 4);
        for (int i = 0; i < iters; i++) {
            Node* t = make_tree(d);
            total += check_tree(t);
            free_tree(t);
        }
    }
    printf("%lld\n", total);
}

File: benchmarks/binary_trees.py
def make_tree(depth):
    if depth == 0: return (1, None, None)
    return (1, make_tree(depth-1), make_tree(depth-1))
def check_tree(node):
    if node[1] is None: return 1
    return 1 + check_tree(node[1]) + check_tree(node[2])
max_depth = 12
total = 0
d = 4
while d <= max_depth:
    iters = 2 ** (max_depth - d + 4)
    for _ in range(iters):
        total += check_tree(make_tree(d))
    d += 2
print(total)
